File: song-convertor.py
import requests

#define api keys
spotify_token = 'SPOTIFY_TOKEN'
youtube_api_key = 'YOUTUBE_API_KEY'



def get_spotify_tracks(playlist_id):
    url = f"https://api.spotify.com/v1/playlists/{playlist_id}/tracks"
    headers = {
        "Authorization": f"Bearer {spotify_token}"
    }
    response = requests.get(url, headers=headers)
    tracks = response.json()['items']
    
    track_list = []
    for item in tracks:
        track_name = item['track']['name']
        artist_name = item['track']['artists'][0]['name']
        track_list.append(f"{track_name} {artist_name}")
    return track_list


def search_youtube(track_name):
    url = f"https://www.googleapis.com/youtube/v3/search?part=snippet&q={track_name}&key={youtube_api_key}&type=video"
    response = requests.get(url)
    items = response.json().get("items", [])
    if items:
        return items[0]["id"]["videoId"]
    return None


def create_youtube_playlist():
   
    pass


def spotify_to_youtube(spotify_playlist_id):
    track_list = get_spotify_tracks(spotify_playlist_id)
    youtube_links = []
    
    for track in track_list:
        video_id = search_youtube(track)
        if video_id:
            youtube_links.append(f"https://www.youtube.com/watch?v={video_id}")
    
 

    return youtube_links

File: song-convertor.js
const SpotifyWebApi = require('spotify-web-api-node');
const { google } = require('googleapis');

// spotify api
const spotifyApi = new SpotifyWebApi({
  clientId: '',
  clientSecret: '',
});

// 
async function authenticateSpotify() {
  const data = await spotifyApi.clientCredentialsGrant();
  spotifyApi.setAccessToken(data.body['access_token']);
}

// import playlists
async function getSpotifyTracks(playlistId) {
  await authenticateSpotify();
  const data = await spotifyApi.getPlaylistTracks(playlistId);
  return data.body.items.map(item => ({
    name: item.track.name,
    artist: item.track.artists[0].name
  }));
}

// youtube api
const youtube = google.youtube({
  version: 'v3',
  auth: '' 
});

//search track on youtub e
async function searchYouTube(track) {
  const res = await youtube.search.list({
    part: 'snippet',
    q: `${track.name} ${track.artist}`,
    maxResults: 1,
    type: 'video'
  });
  return res.data.items[0]?.id?.videoId || null;
}


async function createYouTubePlaylist(authClient) {
  const res = await youtube.playlists.insert({
    part: 'snippet,status',
    auth: authClient,
    requestBody: {
      snippet: {
        title: 'Converted Spotify Playlist',
        description: 'Playlist converted from Spotify'
      },
      status: {
        privacyStatus: 'public'
      }
    }
  });
  return res.data.id;
}


async function addVideoToPlaylist(authClient, playlistId, videoId) {
  await youtube.playlistItems.insert({
    part: 'snippet',
    auth: authClient,
    requestBody: {
      snippet: {
        playlistId: playlistId,
        resourceId: {
          videoId: videoId,
          kind: 'youtube#video'
        }
      }
    }
  });
}


async function spotifyToYouTube(spotifyPlaylistId, authClient) {
  const tracks = await getSpotifyTracks(spotifyPlaylistId);
  const youtubePlaylistId = await createYouTubePlaylist(authClient);

  for (const track of tracks) {
    const videoId = await searchYouTube(track);
    if (videoId) {
      await addVideoToPlaylist(authClient, youtubePlaylistId, videoId);
      console.log(`Added ${track.name} by ${track.artist} to YouTube playlist.`);
    } else {
      console.log(`Could not find ${track.name} by ${track.artist} on YouTube.`);
    }
  }
}


